Redux Toolkit Migration

ChrisMart21 opened the pull request:
# Description

This pull request primarily focuses on updating the project to modern Redux patterns by migrating to the recommended '[Redux-Toolkit](https://redux-toolkit.js.org/)' while addressing several other minor issues along the way.

Fixes:
- #817 
  - Reworked the snippets in question to use react-routerv6's outlets + redux state.
- #92 
  - Data Fetching now managed via RTK Query. Fetching/caching simplified and more robust.
- #886 
  - Preferences removed the container, and added hooks + selectors
- #893 
  - All components now are functional. Dropped the container+component HOC pattern.
- #466 
  - Graphs now scale / are responsive 

Partially addresses 

- #726 / #888 
  -  Continues the ongoing push towards functional components.
  - All graph types other than compare
  - main containers
  - admin preferences/userpages
  
- #1093 
  - Continue ongoing switch towards notifications. 

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations
- This pull request introduces numerous changes across the front end that are yet to be thoroughly tested. While not entirely feature-complete, it represents a significant step towards migrating to RTK. A thorough evaluation is necessary to ensure these changes have not impaired the functionality of the app.
  - Close evaluation and iterations have been ongoing since initial posting of PR. It is reaching its final stages, and is believed to be stable. Further testing and evaluation from the community is always welcome, and encouraged.

- This PR has a version bump to react-router-v6. This change introduces several newer hooks that pair well with the ongoing migration towards functional components. This did however introduce some backwards compatibility issues with the current implementation of the unsaved warning. This component has to be reworked for compatibility with react-router-v6, but also Redux RTK.


 - There was issues/difficulty migrating the entirety of 'Maps'. Maps currently is still using legacy redux and either needs to be reworked, or migrated properly by someone more knowledgeable on this section of the code base.
   - While maps utilizes some newer RTK elements, it is still heavily reliant on legacy redux patterns.  
   - Further work is necessary on this section of the migration.

- Redux-Toolkit introduces some new devmode checks which conflict with some of the legacy Redux code. For example, the 'serializableCheck' has to currently be suppressed due to the use of TimeInverval in Redux State. TimeInterval is a class which is non-serialiazable which needs to be addressed. Additionally 'immutableCheck' is also currently has conflicts with the maps implementation. For the time being, these checks are disabled to prevent the app from malfunctioning, however the goal is to address these issues, and have all dev mode checks enabled. 
  - Progress has been made on this front and will likely be resolved in tandem with the Maps rework.

ChrisMart21 committed:
rtk install

ChrisMart21 committed:
RTK Query Sample Usage

ChrisMart21 committed:
Initial Graph Slice Reducer Refactor

ChrisMart21 committed:
Options Slice Refactor

ChrisMart21 committed:
Current User Slice Refactor

ChrisMart21 committed:
unSavedWarningSliceRefactor

ChrisMart21 committed:
Meter Slice Refactor

ChrisMart21 committed:
Incremental RTK Adoptions

ChrisMart21 committed:
wip

ChrisMart21 committed:
Fix Groups Query for deep meters

ChrisMart21 committed:
fix RPDSelect

ChrisMart21 committed:
wip

ChrisMart21 committed:
Initial ChartDataSelector Refactor

ChrisMart21 committed:
Incremental Changes
--Selects working with refactored & memo'd Chart DataSelectors

ChrisMart21 committed:
Mutations React Router Bump

ChrisMart21 committed:
Initialization Component Changes
- More RTK Queries.

ChrisMart21 committed:
RouteComponentWip
-RouteComponent
- Login mutations, and useNavigateHook();

ChrisMart21 committed:
useSelector changes
- Admin pages Meter, Group, Unit useSelectorChanges.

ChrisMart21 committed:
Routing Changes to Admin and Role Routes.

ChrisMart21 committed:
Ui Options Component Refactor

ChrisMart21 committed:
Line & Bar  Component refactor
- Containers converted to components and utilize RTK Query

ChrisMart21 committed:
Merge remote-tracking branch 'huss/development'

ChrisMart21 committed:
UserChanges.

ChrisMart21 committed:
Merge remote-tracking 'upstream/development'

ChrisMart21 committed:
Query updates

ChrisMart21 committed:
RTK 5.0.0-beta.2 Selector Refactor
- Selectors live in CreateSlice's selector's property

ChrisMart21 committed:
Query Updates, and Data Selectors
-- Redux 5.0.3 Beta version bump
-- Selectors Refactor
-- Custom Cache Behavior for network efficiency

ChrisMart21 committed:
LoadingHook

ChrisMart21 committed:
Fix Merge property usage

ChrisMart21 committed:
Basic Graph State History

ChrisMart21 committed:
Chart Data Select Update

ChrisMart21 committed:
Routing & Unsaved Changes

ChrisMart21 committed:
Add ESLint for React-Hooks

ChrisMart21 committed:
Add Translation TS Completions

ChrisMart21 committed:
Drop Meter, Group, and Unit Slices/Reducer
- All data will now live in API reducers
- Many Misc Changes to admin pages.

ChrisMart21 committed:
RTK Beta Version Bump

ChrisMart21 committed:
Simplify Redux Store
- Export component utilizes query data instead of redux state

ChrisMart21 committed:
Convert Compare to RTKQuery
- Incrementally Delete No-Longer Needed Query Thunks

ChrisMart21 committed:
Sweep for bugs Add TODOS

ChrisMart21 committed:
Api Selector Refactor
- Selectors now wrapped to avoid destructuring assignment across the app

ChrisMart21 committed:
Refactor Query Arg Selectors

ChrisMart21 committed:
Refactor Query Arg Selectors

ChrisMart21 committed:
Separate GraphHistory from Graph
- Update Middleware to mediate between AppState, and Graph

ChrisMart21 committed:
Sweep Selectors for clarity/usage

ChrisMart21 committed:
PRSTASH

ChrisMart21 committed:
Entity Adapter Reafactor

ChrisMart21 committed:
Hist

ChrisMart21 committed:
CreateTunkSlice as init sequence

ChrisMart21 committed:
CutUnsaved warning for maps

ChrisMart21 committed:
Multi Select Changes
- Fix Remove Values.
- Add Disabled indicator w/ tooltip

kaito1105 committed:
test B5

Co-authored-by: Maryam Bouamama <bmamar2005@gmail.com>
Co-authored-by: Sebastian Ramos <seramos@csumb.edu>

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into rtk

ChrisMart21 committed:
Legacy Router Cleanup

ChrisMart21 committed:
Simplify History

ChrisMart21 committed:
Purge Duped Files

ChrisMart21 committed:
Cleanup Legacy code
- Rename slices
- Mark files for deletion

ChrisMart21 committed:
Fix Headers

ChrisMart21 committed:
Fix failed tests

huss commented on src/client/app/redux/api/userApi.ts:
```diff
@@ -0,0 +1,53 @@
+	})
+})
+
+// public async editUsers(users: User[]) {
```
Since these capabilities are above, is this commented out code needed?

ChrisMart21 replied:
Was used as translation reference. This & import at the top needs to get deleted. Will mark as resolved when deleted,

huss commented on src/client/app/redux/api/userApi.ts:
```diff
@@ -0,0 +1,53 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { NewUser, User } from '../../types/items';
+// import { authApi } from './authApi';
```
Is auth done here?

ChrisMart21 replied:
Auth is handled on startup, or login. If authenticated, these request headers\s will have a token for additional verification if needed. That import isn't necessary here.

huss commented on src/client/app/components/admin/UsersDetailComponent.tsx:
```diff
@@ -2,78 +2,79 @@
+		// make new list from existing local user state
+		const updatedList = localUsersChanges.map(user => (user.email === targetUser.email) ? updatedUser : user)
+		setLocalUsersChanges(updatedList)
+		// editUser(user.email, target.value as UserRole);
```
I think this is done in submit so commented out code can be removed.

ChrisMart21 replied:
Agreed.

huss commented on src/client/app/redux/actions/admin.ts:
```diff
@@ -0,0 +1,192 @@
+// import { updateSelectedLanguage } from './options';
+import { graphSlice } from '../slices/graphSlice';
+import { adminSlice } from '../slices/adminSlice';
+// TODO Marked For Deletion after RTK migration solidified
```
Still needed?

ChrisMart21 replied:
This Entire File could be purged. 
It is not used anywhere and was used as reference for translation.

huss commented on src/client/app/redux/actions/admin.ts:
```diff
@@ -0,0 +1,192 @@
+			}
+			if (preferences.defaultLanguage !== state.options.selectedLanguage) {
+				// if the site default differs from the selected language, update the selected language and the locale
+				// dispatch2(updateSelectedLanguage(preferences.defaultLanguage));
```
Is this now done in the component so commented out code no longer needed? Import can go if this goes.

ChrisMart21 replied:
This Entire File could be purged.
It is not used anywhere and was used as reference for translation.

ChrisMart21 replied:
I added a comment to files that need to get purged.

// TODO Marked For Deletion after RTK migration solidified

huss commented on src/client/app/redux/actions/admin.ts:
```diff
@@ -0,0 +1,192 @@
+		const preferences = await preferencesApi.getPreferences();
+		dispatch(adminSlice.actions.receivePreferences(preferences));
+		moment.locale(getState().admin.defaultLanguage);
+		//  TODO reference only DELETE ME
```
Remove?

huss commented on src/client/app/redux/actions/admin.ts:
```diff
@@ -0,0 +1,192 @@
+	};
+}
+
+export function submitPreferencesIfNeeded(): Thunk {
```
I don't see where this function is used. Can you help me understand? (seems shouldSubmitPreferenceData is only used here). This may apply to other, similar functions.

huss commented on src/client/app/components/conversion/ConversionViewComponent.tsx:
```diff
@@ -36,7 +37,7 @@ export default function ConversionViewComponent(props: ConversionViewComponentPr
 		setShowEditModal(false);
 	}
-
+	React.useEffect(() => undefined, [props.conversion])
```
Could a comment be added about what this is doing? I think it updates this modal if the conversion changes.

ChrisMart21 replied:
This should be deleted.

huss commented on src/client/app/components/conversion/ConversionsDetailComponent.tsx:
```diff
@@ -25,23 +20,22 @@ import { Dispatch } from 'types/redux/actions';
+			unitDataById: data && unitsAdapter.getSelectors().selectEntities(data)
+		})
+	})
+	// const x = useAppSelector(state => conversionsApi.endpoints.refresh.select()(state))
```
status?

ChrisMart21 replied:
refreshing now occurs as an onQueryStarted Property of the relevant mutations. Comment can be deleted.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -37,8 +35,9 @@ export default class MetersCSVUploadComponent extends React.Component<MetersCSVU
 		// Refetch meters details.
-		store.getState().meters.hasBeenFetchedOnce = false;
-		fetchMetersDetails();
+		// Removed with rtk migration
```
The meters are not being updated as desired.

huss replied:
fixed

huss commented on src/client/app/components/groups/EditGroupModalComponent.tsx:
```diff
@@ -324,10 +324,10 @@ export default function EditGroupModalComponent(props: EditGroupModalComponentPr
 					// changes were made to the children. This avoid a reload on name change, etc.
-					dispatch(submitGroupEdits(submitState, (i === groupsChanged.length ? true : false) && (childMeterChanges || childGroupChanges)));
-					i++;
+					submitGroupEdits(submitState)
```
Chris?: How does this make sure that all state is okay without reload. Also, since there is an invalidate in src/client/app/redux/api/groupsApi.ts does this do it more than needed?

Maybe look at editUnitMutation for passing parameters.

ChrisMart21 replied:
This and most other admin pages are written such that the page will reset if the cached values it relies on gets modified. In this case the groups cache is updated, therefore resets the current page/ modal. 

It seems however that this mutation is missing the onQueryStarted Logic. The window reload will be replaced with the appropriate cache invalidation(s).

huss replied:
Now that I'm looking at the API, I'm still unsure if this updates too much. The old code determined if other groups were impacted. The new API seems to always invalidate the GroupData. Not sure and other issues with groups.

huss replied:
It is true that this invalidates the entire group redux state. However, it is possible that the edit of one group impacts another that includes it. OED could check if the full update is needed but that would require changing how group state is acquired. Given this isn't done very often and only by admins, OED will not make this change now (or maybe ever). This same philosophy will be used for all such changes/invalidations for admin work.

huss commented on src/client/app/components/AppLayout.tsx:
```diff
@@ -0,0 +1,28 @@
+		<>
+			<ToastContainer transition={Slide} />
+			<HeaderComponent />
+			{props.children ?? <Outlet />}
```
Maybe comment to describe

ChrisMart21 replied:
Sure. Done!

huss commented on src/client/app/components/MapChartComponent.tsx:
```diff
@@ -0,0 +1,387 @@
+		}]
+	};
+
+	/***
```
Can this be removed?

huss replied:
done

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -0,0 +1,102 @@
+						dispatch(groupsApi.util.updateQueryData('getGroups', undefined, groupDataById => { groupsAdapter.updateMany(groupDataById, updates) }))
+					}
+				} catch (e) {
+					console.log(e)
```
This should be changed. log in standatd OED way and/or notify user.

huss replied:
Changed to error popup.

huss commented on src/client/app/components/router/GraphLinkComponent.tsx:
```diff
@@ -0,0 +1,113 @@
+					dispatchQueue.push(appStateSlice.actions.setOptionsVisibility(value === 'true' ? true : false))
+					break;
+				case 'mapID':
+					// 'TODO, Verify Behavior & FIXME! MapLink not working as expected
```
Resolve. Maybe part of map issues.

huss replied:
gone, map link elsewhere

huss committed:
format, console.log removed, simple edits

huss reviewed (commented):
@ChrisMart21 This is not a formal review but some comments for us to work on as I continue to do the formal review. Many will be deleted and are informal.

huss committed:
use Toast msgs + other

- Adds warn and info Toast msgs.
- Removes window.alert.
- Replaces notifyUser and removes from code.
- Does not address window.confirm.
- Some basic edits to CreateMeterModalComponent.

huss commented on src/client/app/components/meters/MetersDetailComponent.tsx:
```diff
@@ -4,129 +4,71 @@
+				{
 					<div className="card-container">
 						{/* Create a MeterViewComponent for each MeterData in Meters State after sorting by identifier */}
+						{/*  Optional Chaining to prevent from crashing upon startup race conditions*/}
```
Can you help me understand why this is optional chaining?

ChrisMart21 replied:
This comment should be removed. It seems that this comment is still lingering around after some of the logic was shifted elsewhere.. This optional chaining still occurs in many of the selectors. 
One issue I was encountering during was that selectors were firing before the initial data was finished loading. This would cause fatal undefined index access issues. To prevent this I utilized optional chaining to short-circuit some of the methods. Doing this caused the methods to gracefully finish and return no data. Not sure if this is ideal, but works for now.

Similarly, the comment above no longer applies either. The sorting of meters is now handled at the store level by entity adapter. This comment should be edited . Data is no longer sorted during the render.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -103,15 +103,16 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
 	const shouldUpdateUnit = (): boolean => {
 		// true if inputted values are okay and there are changes.
 		let inputOk = true;
+		const meterDataByID = selectMeterDataById(store.getState())
```
Curious why this one uses store directly and not useAppSelector as in somw other places.

ChrisMart21 replied:
Good catch! This is bad practice!! It is higly advised against to import store.ts anywhere other than index. Honeslty not sure how this one made it through... Definitely needs to be addressed

huss committed:
basic edits

Does components for meter, router & unit and componentHooks.

huss commented on src/client/app/containers/CompareChartContainer.ts:
```diff
@@ -180,13 +203,15 @@ function mapStateToProps(state: State, ownProps: CompareChartContainerProps): an
 				showgrid: true,
-				gridcolor: '#ddd'
+				gridcolor: '#ddd',
+				fixedrange: true
```
Want to check why added fixedrange?

ChrisMart21 replied:
I believe to avoid zooming when dragging across the graph. If this is desired behavior it can be omitted. Let me know.

huss replied:
Just noting that we don't want to zoom on compare so this is fine.

huss commented on src/client/app/containers/MapChartContainer.ts:
```diff
@@ -339,14 +349,10 @@ function mapStateToProps(state: State) {
-		}
-	};
-	props.config.locale = state.options.selectedLanguage;
+		layout
```
I think removing the locales is why plotly is no longer using the chose language to label the axis items.

ChrisMart21 replied:
Indeed, Corrected the issue.

ChrisMart21 committed:
Auth Midleware, address comments.

ChrisMart21 committed:
Line ErrorBar Fix & Layout Tweaks
- Layout Tweaks to allow for proper graph scaling with window.
- Line utilizes query selectFromResult to derive data in a memoized fashion which should improve performace

huss commented on src/client/app/components/BarChartComponent.tsx:
```diff
@@ -0,0 +1,257 @@
+			{`${translate('select.meter.group')}`}
+		</h1>
+	} else if (!enoughData) {
+		// This normal so plot.
```
First, this does not seem a normal plot as it has no data. Second, I think we may want to change the translate key since not just 3D. Let me know and I can do if you want.

ChrisMart21 replied:
Yes please fix the translations to whatever you find suitable. Also, yes the comment is in the wrong location.

huss replied:
done

huss commented on src/client/app/components/BarChartComponent.tsx:
```diff
@@ -0,0 +1,257 @@
+				useResizeHandler={true}
+				config={{
+					displayModeBar: false,
+					responsive: true
```
locales is gone and does not seem to show language aware labels.

ChrisMart21 replied:
Agreed & fixed

huss commented on src/client/app/components/BarChartComponent.tsx:
```diff
@@ -0,0 +1,257 @@
+					},
+					xaxis: {
+						rangeslider: { visible: true },
+						// rangeselector: { visible: true },
```
Needed?

ChrisMart21 replied:
Not sure. Let me know. I'm still testing things & potential ui controls. I removed the redraw and restore, which i think should come back in some way. Similarly, I added a range slider to bar ,but not entirely sure if its here to stay. We need to solidify the UI design In the near future.

ChrisMart21 replied:
Fixed For now.

huss commented on src/client/app/components/ChartDataSelectComponentSave.tsx:
```diff
@@ -0,0 +1,751 @@
+// /* This Source Code Form is subject to the terms of the Mozilla Public
```
Can this go?

ChrisMart21 replied:
yes! fixed

ChrisMart21 replied:
Yes, Fixed!

huss commented on src/client/app/components/ChartSelectComponent.tsx:
```diff
@@ -3,85 +3,62 @@
-					</DropdownItem>
+					{
+						// Make items for dropdown from enum
+						Object.values(ChartTypes)
```
I like the new code but the items should be alphabetical.

ChrisMart21 replied:
I added a .sort() to the map method. Let me know if this works for you. Otherwise we can rollback.

huss replied:
looks good

ChrisMart21 committed:
Plotly Bar SelectFromResult Refactor

ChrisMart21 committed:
Minor Fixes / Tweaks

huss commented on src/client/app/components/DateRangeComponent.tsx:
```diff
@@ -2,63 +2,52 @@
+import TooltipMarkerComponent from './TooltipMarkerComponent';
+import { selectSelectedLanguage } from '../redux/slices/appStateSlice';
+
+// Potential Fixes, for now omitted
```
Could you explain what the fix is?

ChrisMart21 replied:
There's an issue with the date range picker overflowing & getting overshadowed by the plotly graph when the screen size is not wide enough.

 By default the date picker does not honor the width of the parents container. Enabling this behavior with those few lines of CSS fixes this issue, but I'm not too keen on how it looks visually.

When a chart is displaying, shrink the width of the window some (with and without the css) to see what I mean.

huss replied:
I added a small comment to document.

huss commented on src/client/app/components/ExportComponent.tsx:
```diff
@@ -2,213 +2,186 @@
+					// Divide areaScaling into the rate so have complete scaling factor for readings.
+					const scaling = rateScaling / areaScaling;
+					// Make sure they are there and not being fetched.
+					if (readingsData) {
```
Should this have as lineMeterIsFetching?

ChrisMart21 replied:
Line 66 should cover this, but the other cases should also be carefully evaluated. I suppose a check that there is actually meaningful data like readings.length ... etc could be checked here as well. 

I didn't significantly alter the logic in this or most components,  except for how the data is retrieved. 
Upon careful inspection of this code, it appears that many of these checks can be deferred to some already existing selectors, and possibly re-written for easier maintenance which seams like a near-future endeavor.
A refactor should be in order for this and probably many other files.

huss committed:
fix export bug + edits

The test for meters had an issue for convertToCSV.

huss commented on src/client/app/components/GraphicRateMenuComponent.tsx:
```diff
@@ -22,13 +23,10 @@ export default function GraphicRateMenuComponent() {
 	const dispatch = useDispatch();
```
Most places use useAppDispatch but a few use the old useDispatch. Wondering why?

ChrisMart21 replied:
The behavior 'should' be identical & not cause issues, however they all really should be using 'useAppDispatch'.  Must've been overlooked during translation.

huss replied:
I replaced all uses with useAppDispatch.

huss commented on src/client/app/components/HeaderButtonsComponent.tsx:
```diff
@@ -77,15 +76,16 @@ export default function HeaderButtonsComponent() {
+	const { profile: currentUser } = useAppSelector(selectCurrentUser);
 	// Tracks unsaved changes.
-	const unsavedChangesState = useSelector((state: State) => state.unsavedWarning.hasUnsavedChanges);
+	// TODO Re-implement AFTER RTK Migration
```
Is unsaved still used and status of this comment? Does it need to be done later per TODO below?

ChrisMart21 replied:
Unsaved warning needs an overall update and is planned to get re-implemented post-rtk migration.

huss commented on src/client/app/components/HeaderComponent.tsx:
```diff
@@ -3,36 +3,24 @@
-		display: 'inline-block',
-		marginTop: '10px'
-	};
+	const siteTitle = useAppSelector(state => state.admin.displayTitle);
```
Another palce where admin state is directly accessed.

ChrisMart21 replied:
Corrected.

huss commented on src/client/app/components/HistoryComponent.tsx:
```diff
@@ -0,0 +1,35 @@
+
+	return (
+		<div style={{ width: '80%', display: 'flex', flexDirection: 'row', justifyContent: 'space-between' }}>
+			<svg width={20} height={20} viewBox="0 0 10 10" fill="none" xmlns="http://www.w3.org/2000/svg"
```
xmlns="http://www.w3.org/2000/svg" does not seem necessary when I removed it. Should we have this URL to w3?

huss commented on src/client/app/components/LineChartComponent.tsx:
```diff
@@ -0,0 +1,152 @@
+import locales from '../types/locales';
+
+// Stable reference for when there is not data.
+const stableEmptyReadings: LineReadings = {}
```
Is it okay to move this inside the function? It only seems to be used here but maybe there is an efficienty issue I'm not thinking of.

ChrisMart21 replied:
It is intentionally left outside. Since it is declared outside it will retain a stable object reference. This is there to prevent unnecessary re-rendering when initial data is in flight. Putting it inside will create a new reference each render which defeats its purpose. These stable references could be moved to a different location , or maybe wrapped on a selector if desired. This pattern will show up in a handful of places so we need to decide on this sooner than later.

huss replied:
Clarified comment

huss committed:
login translation, remove spinner commented out code & basic edits

huss commented on src/client/app/components/ReadingsPerDaySelectComponent.tsx:
```diff
@@ -2,58 +2,45 @@
-	})
+	}
+
+	// Return normal interval
```
First, could this commented out code be removed? Also, I'm unclear on why the new code returns the same value as the old code in this case. Finally, should there be a return above and not be implicit?

ChrisMart21 replied:
Commented out code, removed.

ChrisMart21 replied:
What do you mean by old code. vs new code. I'm unsure there's a difference between the two iterations in this context. Could you expand on this?

ChrisMart21 committed:
ChartLink bug fix + single dispatch
- 3d Chart Alignment Fix

huss committed:
basic edits +

Also remove unused props, fix JSDoc, TODO, unsaved translation

huss committed:
Merge remote-tracking branch 'ChrisMart21/rtk' into pr/ChrisMart21/1113

ChrisMart21 committed:
Select-Tooltip fix.

huss commented on src/client/app/redux/actions/conversions.ts:
```diff
@@ -1,75 +1,40 @@
+// @ts-nocheck
+/* eslint-disable jsdoc/require-param */
+
+// TODO Marked For Deletion after RTK migration solidified
```
Not sure used and checking if you are okay with removing this now?

ChrisMart21 replied:
Yep, Should go.

huss commented on src/client/app/redux/actions/graph.ts:
```diff
@@ -0,0 +1,30 @@
+import * as t from '../../types/redux/graph';
+import { ComparePeriod, SortingOrder } from '../../utils/calculateCompare';
+
+export interface LinkOptions {
```
I don't see this used. Ok to remove?

ChrisMart21 replied:
yes, this can/should go

huss commented on src/client/app/redux/actions/groups.ts:
```diff
@@ -0,0 +1,193 @@
+//  * License, v. 2.0. If a copy of the MPL was not distributed with this
+//  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// TODO Marked For Deletion after RTK migration solidified
```
Ok to remove?

huss replied:
done

huss commented on src/client/app/redux/actions/meters.ts:
```diff
@@ -1,65 +1,41 @@
-	return { type: ActionType.ReceiveMetersDetails, data };
-}
 
 export function fetchMetersDetails(): Thunk {
```
This function is still in a comment in src/client/app/components/csv/ReadingsCSVUploadComponent.tsx but maybe it can be cleaned up and removed. The rest seem to be unused so can delete per TODO above.

huss replied:
done

huss commented on src/client/app/redux/actions/units.ts:
```diff
@@ -1,61 +1,41 @@
-import * as t from '../types/redux/units';
-import { unitsApi } from '../utils/api';
-import { updateCikAndDBViewsIfNeeded } from './admin';
+// TODO Marked For Deletion after RTK migration solidified
```
I don't think this is used. Ok to remove?

huss replied:
done

huss commented on src/client/app/redux/api/authApi.ts:
```diff
@@ -0,0 +1,60 @@
+			// in this case, a user logged in which means that some info for ADMIN meters groups etc.
+			// invalidate forces a refetch to any subscribed components or the next query.
+			invalidatesTags: ['MeterData', 'GroupData']
+			// Listeners for this query (ExtraReducers):
```
What is this commented out code for?

ChrisMart21 replied:
RTK Slices have an optional property called extra reducers, which can listen to any action defined outside of itself. Sometimes its difficult to know which slices are listening to things like rtkQuery actions. In this case its current user slice listens for the match fulfilled action. This was an early on idea- that I'm not certain this will work cleanly. It wasto document the listeners for the endpoint, but should probably go.

ChrisMart21 replied:
removing for now since not standard.

huss commented on src/client/app/redux/api/baseApi.ts:
```diff
@@ -0,0 +1,42 @@
+	],
+	// Initially no defined endpoints, Use rtk query's injectEndpoints
+	endpoints: () => ({})
+	// Defaults to 60 seconds or 1 minute
```
Keep commented out code?

ChrisMart21 replied:
Deleted, can change if needed.

huss commented on src/client/app/redux/api/metersApi.ts:
```diff
@@ -0,0 +1,114 @@
+
+
+export const selectMeterDataResult = metersApi.endpoints.getMeters.select()
+
```
Is it standard practice to export even if only used in this file? Just trying to understand.

ChrisMart21 replied:
Not sure if there's a standard practice here, or if there are performance implications. Just think its useful to have selectors accessible wherever if need.

huss replied:
Maye see https://stackoverflow.com/questions/76705375/does-exporting-a-module-and-never-importing-it-anywhere-affect-performance

huss replied:
Maybe see https://stackoverflow.com/questions/76705375/does-exporting-a-module-and-never-importing-it-anywhere-affect-performance. I'm not sure which way to go on this.

ChrisMart21 replied:
I think this is a non-critical issue. I will deffer this to some later PR if desired.

huss commented on src/client/app/redux/api/readingsApi.ts:
```diff
@@ -0,0 +1,156 @@
+			// Refer to the line endpoint for an example of an endpoint with custom cache behavior
+			query: ({ id, timeInterval, graphicUnitId, readingInterval, meterOrGroup }) => ({
+				// destructure args that are passed into the callback, and generate/ implicitly return the API url for the request.
+				url: `api/unitReadings/threeD/${meterOrGroup}/${id}?`,
```
Should there be a ? at the end (below does not have)?

huss replied:
Should there be a ? at the

ChrisMart21 replied:
Interesting that it works either way. I'm assuming the fetchBaseQuery is smart enough to sort itself out. I guess it should be omitted since it handles the url generation internally when given a url + params object.

huss replied:
Removed.

huss committed:
basic edits +

Also make error logic consistent in readingsApi.ts

huss commented on src/client/app/redux/componentHooks.ts:
```diff
@@ -0,0 +1,38 @@
+// usage
+// const translate = useTranslate()
+// translate('myKey')
+export const useTranslate = () => {
```
I'm trying to figure out how this relates to src/client/app/utils/translate.ts and the fact that some places use one or the other.

huss replied:
Need to resolve store more generally so will address in future.

huss committed:
basic edits

ChrisMart21 committed:
ChartLink Component rework.

huss committed:
fix up packages

This puts back previous package changes and only changes the ones desired for this PR.

huss committed:
Merge remote-tracking branch 'ChrisMart21/rtk' into pr/ChrisMart21/1113

huss commented on src/client/app/redux/selectors/adminSelectors.ts:
```diff
@@ -0,0 +1,309 @@
+			}
+		});
+		// Put in alphabetical order.
+		possibleMeterUnits = new Set(_.sortBy(Array.from(possibleMeterUnits), unit => unit.identifier.toLowerCase(), 'asc'));
```
This is not a change but I noticed in src/client/app/redux/api/unitsApi.ts that the sorting includes an accent term and localCompare. There are other places where the way we sort differs. Do you think they should be consistent? If so, we should see if a function(s) can be used to do this in an easier way.

huss replied:
Will get consistent in future.

huss commented on src/client/app/redux/selectors/adminSelectors.ts:
```diff
@@ -0,0 +1,309 @@
+		// Graphic units incompatible with currently selected unit
+		const incompatibleGraphicUnits = new Set<UnitData>();
+		// If unit is not 'no unit'
+		if (unitId != -99) {
```
This is used for both edit and create. For create there used to be a separate test for -999. I'm noting since this may need to change to get the other code to allow no unit. Same below.

huss replied:
I think this is resolved. Could @ChrisMart21 verify.

ChrisMart21 replied:
Is resolved with recent create meter modal updates.

huss commented on src/client/app/redux/selectors/adminSelectors.ts:
```diff
@@ -0,0 +1,309 @@
+			gps: '',
+			// Defaults of -999 (not to be confused with -99 which is no unit)
+			// Purely for allowing the default select to be "select a ..."
+			unitId: -99,
```
-99 vs -999 here too.

huss replied:
Changes have fixed.

huss commented on src/client/app/redux/selectors/threeDSelectors.ts:
```diff
@@ -0,0 +1,47 @@
+	(id, meterOrGroup, meterDataById, groupDataById) => {
+		//Default Values
+		let meterOrGroupName = 'Unselected Meter or Group'
+		let isAreaCompatible = true;
```
My reading of the code indicates this must be set by selectNameFromEntity so this value does not matter. Wanted to verify that is correct. If so, I think we should set to empty string (or comment).

ChrisMart21 replied:
This is correct. Not is an empty string instead.

huss commented on src/client/app/redux/selectors/uiSelectors.ts:
```diff
@@ -0,0 +1,545 @@
+		// and not shift the graphs x-axis so using.
+		return new TimeInterval(moment(newMinXTimestamp), moment(newMaxXTimestamp)).toString();
+	} else {
+		return 'all'
```
This used to throw an error. I can see that all would graph the full range but is it covering up an error? Should we log here or is it normal?

ChrisMart21 replied:
I think this approach needs to be reworked. We need to tap into the plotly component callbacks to retrieve and manage this value.
 I haven't got around to implementing this and kept this in as a reminder/to cover up errors for the time being.

ChrisMart21 replied:
This has been pushed back until after this PR. Graphs are going to all be reworked to track all necessary attrributes will live in Redux( including the range Slider interval) Resolving for now.

huss committed:
minor edits/commenting

huss committed:
remove adminState.selectedMeter

This existed before this PR but this state was no longer used.
Might want to look at in the future if add meter dropdown in CSV import.

huss commented on src/client/app/redux/slices/graphSlice.ts:
```diff
@@ -0,0 +1,380 @@
+				isAMeter ? current.selectedMeters = [] : current.selectedGroups = []
+
+			}
+			if (valueRemoved) {
```
Are these mutually exclusive? If so, should it be else if for these?

ChrisMart21 replied:
yes, only one event should happpen per reducer pass. elifs were used in the later half of this reducer, but didn't seem to make it up here. Fixed

huss commented on src/client/app/types/locales.ts:
```diff
@@ -2,7 +2,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-export default class Locales{
+export default class locales{
```
Normally class names are capitalized. Is this to avoid the same name from Plotly or something else? Maybe we should refactor name if this is the case.

ChrisMart21 replied:
Reverted back to UpperCase.

huss commented on src/client/app/types/redux/state.ts:
```diff
@@ -16,15 +15,15 @@ import { CurrentUserState } from './currentUser';
 	readings: {
 		line: LineReadingsState;
 		bar: BarReadingsState;
 		compare: CompareReadingsState;
```
Just curious why 3D was removed from here.

ChrisMart21 replied:
This is old state, which is only used in maps. 3d was the first to implement RTKQ, so was removed here a while ago. But this all should go soon, so not of much importance. Resolving convo.

huss committed:
basic edits

I make all uses of timezone have zone uncaptialized. Removed unused enum ByMeterOrGroup.

huss commented on src/client/app/utils/hasPermissions.ts:
```diff
@@ -21,5 +21,7 @@ export function hasPermissions(user: UserRole, compareTo: UserRole): boolean {
  * @returns Whether or not user is an admin
  */
 export function isRoleAdmin(user: UserRole): boolean {
+	// TODO Already Converted to a selector
```
The only place this is used is src/client/app/components/HeaderButtonsComponent.tsx. The comment seems out of date since that function does not exist. I think we should purge the last use and get rid of this but want to check with you on the best way to do that.

ChrisMart21 replied:
Yeah, these are essentially selectors. This file is legacy, and will be deleted.

huss commented on src/server/test/web/readingsData/expected_bar_ri_15_mu_kWh_gu_kWh_st_-inf_et_inf_bd_75.csv:
```diff
@@ -0,0 +1,2 @@
+reading,start time,end time
```
Not sure why this is listed as a change but same as current development so should be fine. Just noting.

huss commented on src/server/test/web/readingsBarMeterQuantity.js:
```diff
@@ -94,9 +94,23 @@ mocha.describe('readings API', () => {
-
-                // Add B5 here
-
+                mocha.it('B5: 75 day bars for 15 minute reading intervals and quantity units with +-inf start/end time & kWh as kWh', async () => {
```
Another change that already in development so noting.

huss committed:
simple edits

Plus one comment.

huss commented on src/client/app/components/groups/GroupsDetailComponent.tsx:
```diff
@@ -3,77 +3,29 @@
-		// In case the group state was not yet loaded you need to do this again.
-	}, [groupsStateLoaded, groupsAllChildrenLoaded]);
+	// page may contain admin info so verify admin status while admin is authenticated.
+	authApi.useTokenPollQuery(undefined, { skip: !isAdmin, pollingInterval: authPollInterval })
```
I'm wondering why tis is not in admin, conversion & unit (and maps at some point) pages? Maybe others.

ChrisMart21 replied:
Admin outlet has this polling effect so any of it's children also inherit the behavior.

Meter and group pages are kinda different because they modify behaviors based on auth status so they live outside the adminOutlet and explicitly need the polling added here

huss committed:
minor edits +

Added some translations.

ChrisMart21 committed:
Fix Circular

ChrisMart21 committed:
Header

ChrisMart21 committed:
File Restructure.

ChrisMart21 committed:
Merge remote-tracking branch 'upstream/development' into rtk
- Merge / Migrate CIK to RTK

huss committed:
small edits

ChrisMart21 committed:
More Selector File shifts

ChrisMart21 committed:
Bar Export Fix
- Bar was pointing to wrong endpoint ('line')

ChrisMart21 committed:
Initial Export Rework

ChrisMart21 committed:
Export as Thunks Continued

ChrisMart21 committed:
GraphNav Initial

huss committed:
generalize translation key

huss committed:
Merge remote-tracking branch 'ChrisMart21/rtk' into pr/ChrisMart21/1113

ChrisMart21 committed:
Fix Slider Range

ChrisMart21 committed:
PlotlyMinHeight

ChrisMart21 committed:
MorePlotly

huss committed:
address some PR comments

Remove comment, file & unneeded ? in URL.

huss committed:
fix export label +

Now correctly labels groups in export file. Also, basic edit.

huss committed:
enable ESLint semicolon rule

huss committed:
fix TS/cleint files for semicolon

huss committed:
Merge remote-tracking branch 'origin/development' into pr/ChrisMart21/1113

huss committed:
fix MPL headers

huss committed:
fix LG6 merge issue

Two lines of code were lost in the recent merge.
This adds them back. It is the original code from the developer.

huss committed:
remove resolved comment

huss committed:
convert console.log to error notification

huss committed:
add comment on CSS value

huss committed:
useDispatch -> useAppDispatch

huss committed:
Clarify comment

ChrisMart21 committed:
Restore Options Scroll Bars

ChrisMart21 committed:
Restore Options Export

huss committed:
update icons

This removes the Windows identifier file.
It also uses a refresh with MIT license.

ChrisMart21 committed:
Merge branch 'rtk' of github.com:ChrisMart21/OED-CSUMB-Capstone-Sum23 into rtk

ChrisMart21 committed:
RollBack Plotly + Tweaks
- No longer honors, saves, or links to slider range, in favor of short-term stability

ChrisMart21 committed:
Map Chart Select Restore

ChrisMart21 committed:
Purge unused Legacy Api endpoints

ChrisMart21 committed:
Keep maps as is

ChrisMart21 committed:
Min Heigh Plots

huss committed:
remove unneeded await

ChrisMart21 committed:
Fix Maps Data Fetch
- Ui Select Change

ChrisMart21 committed:
Fix Alignment on Admin pages

ChrisMart21 committed:
Create Meter No Unit Fix

ChrisMart21 committed:
Fix Create MeterModal

ChrisMart21 committed:
cik refresh restored.

ChrisMart21 committed:
Radar Include Rate Menu

huss committed:
Merge remote-tracking branch 'ChrisMart21/rtk' into pr/ChrisMart21/1113

ChrisMart21 committed:
Address PR Comments.

ChrisMart21 committed:
Re-home stable emtpy readings

ChrisMart21 committed:
3d Revisions
- rework/tweak Readings Per Day Component modern RTK.

ChrisMart21 committed:
More 3d Revisions
- default to 6 months back when selecting entity to graph on 3d.

ChrisMart21 committed:
3d Bugs

ChrisMart21 committed:
update dispatch 3d from selects.
-- remove logs.

ChrisMart21 committed:
logs

huss committed:
minor format/edit

huss committed:
remove actions now in RTK

radar and 3D files removed since radar always used line and 3D seems similar.
Thus, that state definations are also gone.

huss committed:
defaultGraphicUnit -> unitId bug fix

The selectUnitNmae used the wrong unit id.

ChrisMart21 committed:
createMeterModalTweak

huss committed:
fix bug where meter/group dropdown wrong for maps

huss committed:
update group edit state so correct after create

huss committed:
invalidated readings on appropriate conversion change

huss committed:
comment out/TODO for 3D when > 1 yr

huss committed:
Merge remote-tracking branch 'ChrisMart21/rtk' into pr/ChrisMart21/1113

huss commented:
I want to express my deep thanks to @ChrisMart21 for initiating and completing the migration of OED to the Redux Toolkit and Query system. This not only gets OED up to the latest Redux usage but adds a number of new features such as history of graphics. This is a substantial step in OED keeping current with technology and reducing our technical debt. Many thanks.

huss reviewed (approved):
Thanks to @ChrisMart21 for another substantial contribution. All the comments have been addressed and testing finds the code works as expected. It is ready to merge.
